examples/vision/randaugment.py

```diff
@@ -0,0 +1,296 @@
+"""
+"""
+Data augmentation is a very useful technique that helps to improve the translational
+invariance of convolutional neural networks (CNN). RandAugment is a stochastic data
+augmentation routine for vision data and was proposed in [RandAugment: Practical
+automated data augmentation with a reduced search
+space](https://arxiv.org/abs/1909.13719). It is composed of strong augmentation
+transforms like color jitters, Gaussian blurs, saturations, etc. along with more
+traditional augmentation transforms such as random crops.
+
+RandAugment has two parameters:
+* `n` that denotes the number of randomly selected augmentation transforms to apply
+sequentially
+* `m` strength of all the augmentation transforms
+
+These parameters are tuned for a given dataset and a network architecture. The authors of
+RandAugment also provide pseudocode of RandAugment in the original paper (Figure 2):
+
+![](https://i.ibb.co/Df6Ynxd/image.png)
+
+Recently, it has been a key component of works like [Noisy Student Training](https://arxiv.org/abs/1911.04252) and
+[Unsupervised Data Augmentation for Consistency Training](https://arxiv.org/abs/1904.12848). It has been also central to the
+success of [EfficientNets](https://arxiv.org/abs/1905.11946).
+
+This example requires TensorFlow 2.4 or higher, as well as
+[`imgaug`](https://imgaug.readthedocs.io/), which can be installed using the following
+command:
+"""
+
+"""shell
+pip install -U -q imgaug
+"""
+
+"""
+## Setup
+"""
+import numpy as np
+import tensorflow as tf
+import matplotlib.pyplot as plt
+from tensorflow.keras import layers
+
+import tensorflow_datasets as tfds
+
+tfds.disable_progress_bar()
+
+from imgaug import augmenters as iaa
+import imgaug as ia
+
+"""
+## Load the CIFAR10 dataset
+
+For this example, we will be using the [CIFAR10
+dataset](https://www.cs.toronto.edu/~kriz/cifar.html).
+"""
+
+(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
+print(f"Total training examples: {len(x_train)}")
+print(f"Total test examples: {len(x_test)}")
+
+"""
+## Define hyperparameters
+"""
+
+AUTO = tf.data.AUTOTUNE
+BATCH_SIZE = 128
+EPOCHS = 10
+
+"""
+## Initialize `RandAugment` object
+
+Now, we will initialize a `RandAugment` object from the `imgaug.augmenters` module with
+the parameters suggested by the RandAugment authors.
+"""
+
+rand_aug = iaa.RandAugment(n=3, m=7)
+
+
+def augment(images):
+    # Input to `augment()` is a TensorFlow tensor which
+    # is not supported by `imgaug`. This is why we first
+    # convert it to its `numpy` variant.
+    return rand_aug(images=images.numpy())
+
+
+"""
+## Create TensorFlow `Dataset` objects
+
+There's one problem, though. We cannot map our `augment()` function to a TensorFlow
+`Dataset` object. To tackle this problem, we will make use of the
+[`tf.py_function`](https://www.tensorflow.org/api_docs/python/tf/py_function) that can
+convert a Python function into a TensorFlow op.
```

Comment:
`py_function` is pretty advanced and also very useful, so I think it deserves an extensive introduction.

```
Because `RandAugment` can only process NumPy arrays, it cannot be applied directly as part of the `Dataset` object (which expects TensorFlow tensors). To make `RandAugment` part of the dataset, we need to warp it in a `tf.py_function`.

A `tf.py_function` is a TensorFlow operation (which, like any other TF operation, takes TF tensors as arguments and returns TF tensors) that is capable of running arbitrary Python code. Naturally, this Python code can only be executed on CPU (whereas the rest of the TensorFlow graph can be accelerated on GPU), which in some  cases can cause significant slowdowns -- however, in this case, the `Dataset` pipeline will run asynchronously together with the model, and doing preprocessing on CPU will remain performant.
```

etc. -- you can expand on it: when to use it, how it works, what are the tradeoffs...

Reply:
Absolutely. I will include it. Thank you, Francois!